services/wfm/src/main/java/org/openkilda/wfm/topology/flow/service/FlowService.java

```diff
@@ -0,0 +1,213 @@
+ *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *   See the License for the specific language governing permissions and
+ *   limitations under the License.
+ */
+
+package org.openkilda.wfm.topology.flow.service;
+
+import org.openkilda.model.Flow;
+import org.openkilda.model.FlowPair;
+import org.openkilda.model.FlowPair.FlowPairBuilder;
+import org.openkilda.model.FlowStatus;
+import org.openkilda.model.Switch;
+import org.openkilda.persistence.PersistenceManager;
+import org.openkilda.persistence.TransactionManager;
+import org.openkilda.persistence.repositories.FlowRepository;
+import org.openkilda.persistence.repositories.FlowSegmentRepository;
+import org.openkilda.persistence.repositories.IslRepository;
+import org.openkilda.persistence.repositories.SwitchRepository;
+
+import java.time.Instant;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+public class FlowService {
+    private PersistenceManager persistenceManager;
+    private FlowRepository flowRepository;
+    private FlowSegmentRepository flowSegmentRepository;
+    private SwitchRepository switchRepository;
+    private IslRepository islRepository;
+
+    public FlowService(PersistenceManager persistenceManager) {
+        this.persistenceManager = persistenceManager;
+        flowRepository = persistenceManager.getRepositoryFactory().createFlowRepository();
+        flowSegmentRepository = persistenceManager.getRepositoryFactory().createFlowSegmentRepository();
+        switchRepository = persistenceManager.getRepositoryFactory().createSwitchRepository();
+        islRepository = persistenceManager.getRepositoryFactory().createIslRepository();
+    }
+
+    public Iterable<Flow> getFlow(String flowId) {
+        return flowRepository.findById(flowId);
+    }
+
+    /**
+     * Stores flow and it's segments into DB.
+     *
+     * @param flowPair - forward and reverse flows to be saved.
+     */
+    public void createFlow(FlowPair flowPair) {
+        TransactionManager transactionManager = persistenceManager.getTransactionManager();
+        transactionManager.begin();
+        try {
+            createFlowForPair(flowPair);
+            transactionManager.commit();
+        } catch (Exception e) {
+            transactionManager.rollback();
+            throw e;
+        }
+    }
+
+    private void createFlowForPair(FlowPair flowPair) {
+        Flow flow = flowPair.getForward();
+        flowSegmentRepository.deleteFlowSegments(flow);
+        processCreateFlow(flow);
+        flow = flowPair.getReverse();
+        processCreateFlow(flow);
+    }
+
+    private void processCreateFlow(Flow flow) {
+        Switch srcSwitch = switchRepository.findBySwitchId(flow.getSrcSwitchId());
+        Switch dstSwitch = switchRepository.findBySwitchId(flow.getDestSwitchId());
+        flow.setSrcSwitch(srcSwitch);
```

Comment:
Should it be done by flow repository?

Reply:
not sure on that.

Reply:
Refactored.

Reply:
Refactored. Now it's done in the repository.